### ISSUE TITLE
Tracker: geopandas feature parity

### ISSUE THREAD

**Issue by kylebarron** [kylebarron/geopolars#13]
Make todo list here of geopandas functions and implementation status

**Comment by kylebarron** [comment]
Looks great! As a next step we might want to assign a difficulty to each of these? Like some of the geometric manipulations are quite easy because they just wrap algorithms in `geo`. But spatial joins I assume would be more difficult and require our own implementation or more than minimal binding code.

Unclear whether it should be part of _this list_, (maybe better as another issue) but another priority would be FFI from Python. So that it would only be one or two lines to do
```py
geopolars.GeoDataFrame.from_geopandas(gdf)
```
This would probably use `pyarrow` to construct a `pyarrow.Table` and then wrap/extend [pypolars' FFI](https://github.com/pola-rs/polars/blob/master/py-polars/src/arrow_interop/to_rust.rs) to construct a polars DataFrame. The cool part here is if we end up being able to use Arrow extension types, then we might not even need to extend pypolars at all! Ideally the ffi would preserve the extension type metadata, and the standard `polars.from_arrow(gdf.to_arrow())` would work out of the box.

**Comment by stuartlynn** [comment]
Sounds good about the difficulty. I wonder if we should first break each of these out into their own ticket for discussion and tag with difficulty there. Can link to this list with the ticket ids to track?

Nice about the ffi! I think we should make that it's own ticket though as it's less about reaching feature parity with geopandas and more about being able to interoperate with it.

**Comment by kylebarron** [comment]
agreed on both

**Comment by stuartlynn** [comment]
Cool will start breaking these out in to tickets but leave the ffi ticket for yourself. I think you have more context on it than I do.

**Comment by stuartlynn** [comment]
Added a bunch of these. Tried to include some implementation details and thoughts and links to both the geopandas equivalent and any relevant functions in the geo / other crates.

A couple of quick thoughts here 

- A lot of the Geopandas functions have an inplace option, to either return a new GeoSeries / GeoDataFrame or modify the existing one. Is this something we want to support or just have an immutable style API?
- There are a bunch of things that Geopandas doesn't implement, are we aiming just for a 1-1 mapping of the API or are we open to other functionality? Obviously with the understanding that getting to parity with Geopandas s the priority.
- Geopandas allows there to be different Geometry types in the same GeoSeries (for example [here]() ). It seems like, especially in the case of a dataset that has a combination of say for example MultiPolygons and Polygons , this would be useful to allow in geopolars as well. 

Will finish up the remainder tomorrow. At that point we might want to create some difficulty labels on github and tag each of these issues. I think @kylebarron  you have to do that being the project owner

**Comment by kylebarron** [comment]
> * A lot of the Geopandas functions have an inplace option, to either return a new GeoSeries / GeoDataFrame or modify the existing one. Is this something we want to support or just have an immutable style API?

I don't think so... I think geopandas should be a guide but not entirely dictate the API here. In this case, polars itself seems to have a very strong immutable API: https://github.com/pola-rs/polars/issues/3506#issuecomment-1138263501. And I think Arrow memory is usually immutable so that it can be shared across threads. So that would make me lean towards geopolars also having an immutable API.

> * There are a bunch of things that Geopandas doesn't implement, are we aiming just for a 1-1 mapping of the API or are we open to other functionality? Obviously with the understanding that getting to parity with Geopandas s the priority

I think geopandas parity would be nice. But beyond that; if an algorithm is implemented in `geo` but not in `geopandas`... I don't see a reason not to expose it. It seems like a thin wrapper that wouldn't be difficult to implement or maintain.

> * Geopandas allows there to be different Geometry types in the same GeoSeries (for example here ). It seems like, especially in the case of a dataset that has a combination of say for example MultiPolygons and Polygons , this would be useful to allow in geopolars as well.

IMO this could add significant complexity and deserves its own discussion. For one, the arrow-native encoding doesn't yet define how to store mixed geometry types: https://github.com/geopandas/geo-arrow-spec/blob/main/format.md#mixed-geometry-types
